snapshotter/volume_integ_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/oci"
+	"github.com/firecracker-microvm/firecracker-containerd/internal/integtest"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	"github.com/firecracker-microvm/firecracker-containerd/runtime/firecrackeroci"
+	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/internal/integtest/stargz/fs/source"
+	"github.com/firecracker-microvm/firecracker-containerd/volume"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const mib = 1024 * 1024
+
+func TestGuestVolumeFrom_Isolated(t *testing.T) {
+	integtest.Prepare(t, integtest.WithDefaultNetwork())
+	const (
+		vmID     = "default"
+		postgres = "docker.io/library/postgres:14.3"
+		alpine   = "docker.io/library/alpine:3.10.1"
+		runtime  = "aws.firecracker"
+	)
+
+	ctx := namespaces.WithNamespace(context.Background(), vmID)
+
+	client, err := containerd.New(integtest.ContainerdSockPath, containerd.WithDefaultRuntime(runtime))
+	require.NoError(t, err, "unable to create client to containerd service at %s, is containerd running?", integtest.ContainerdSockPath)
+	defer client.Close()
+	fcClient, err := integtest.NewFCControlClient(integtest.ContainerdSockPath)
+	require.NoError(t, err, "Failed to create fccontrol client")
+
+	vs := volume.NewSet(runtime)
+
+	// Add a non-stargz image with volumes.
+	localImage := volume.FromImage(client, postgres, "postgres-snapshot", volume.WithSnapshotter("devmapper"))
+	err = vs.AddFrom(ctx, localImage)
+	require.NoError(t, err)
+
+	// Add a stargz image.
+	// The volume directories must be specified since the host's containerd doesn't know about the image.
+	remoteImage := volume.FromGuestImage(
+		client, vmID, al2stargz, "al2-snapshot", []string{"/etc/yum"},
+		volume.WithSnapshotter("demux"),
+	)
+	mount, err := vs.PrepareDriveMount(ctx, 10*mib)
+	require.NoError(t, err)
```

Comment:
I cannot do `err = vs.AddFrom(ctx, remoteImage)` here since the image cannot be pulled before starting the VM.

Alternatively we could make volume.Set "smarter" and treat GuestVolumeImageProvider somewhat differently.

Reply:
Thinking through the way you actually setup volumes, we need one of:
`Set.PrepareDirectory`
or
`Set.PrepareDriveMount` and maybe `GuestVolumeImageProvider.Copy`

If we made `volume.Set` smarter so that we could user `vs.AddFrom` here, then it seems like we could make the setup a little less confusion with something like:

```
mount, err := set.PrepareDriveMount(ctx, 10*mib)
// launch vm
err = set.PrepareInGuestVolumes(ctx)
```

Maybe there's value in renaming the `Prepare` methods to match the function they provider rather than what they literally do? e.g.
`set.PrepareHostOnlyVoumes`, `set.PrepareVMPreLaunchVolumes`, `set.PrepareVMPostLaunchVolumes`

A couple things I'm not sure of with this proposal:
1. Does it push too much logic into `volume.Set`? 
2. Is it too obfuscated from what's actually going on (tbf, I'm not sure `set.AddFrom` is setting up temp mounts of the image or that `set.PrepareDriveMount` is copying data into a thin device)

Reply:
Let me think about. volume.Set isn't aware about Providers, but I probably should prioritize usability over extensibility this time.